From 04f1870c91ab7ed2a93467e717e1e7bb97136113 Mon Sep 17 00:00:00 2001
From: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Date: Tue, 30 Apr 2013 19:38:46 +0700
Subject: Dalvik: Handle class initialization failure in dvmGetDeclaredConstructors

BZ: 104701

In case of class initialization failure (e.g. UnsatisfiedLinkerError) we have
to return immediately in order to throw an exception correctly. Otherwise we
can face dvmAbort from findClassNoInit during "Fill out the array" step.

Category: device-enablement
Domain: AOSP-Dalvik-Runtime
Origin: internal
Upstream-Candidate: yes

Change-Id: I498a9ccb76397f465214bc84efb4ac3647aad1da
Orig-MCG-Change-Id: I0b3c9e5ed3145372a8044debd89c9a5f3f6cec7f
Signed-off-by: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/reflect/Reflect.cpp |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/vm/reflect/Reflect.cpp b/vm/reflect/Reflect.cpp
index 9cb9fc7..c97df4e 100644
--- a/vm/reflect/Reflect.cpp
+++ b/vm/reflect/Reflect.cpp
@@ -463,7 +463,12 @@ ArrayObject* dvmGetDeclaredConstructors(ClassObject* clazz, bool publicOnly)
      * We're bypassing the normal resolution mechanism, so we init it here.
      */
     if (!dvmIsClassInitialized(clazz))
-        dvmInitClass(clazz);
+    {
+        if(!dvmInitClass(clazz))
+        {
+            return NULL;
+        }
+    }
 
     /*
      * Count up the #of relevant methods.
-- 
1.7.4.1

